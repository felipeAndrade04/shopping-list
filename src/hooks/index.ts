export * from './use-auth';
export * from './use-app-dispatch';
export * from './use-app-selector';
