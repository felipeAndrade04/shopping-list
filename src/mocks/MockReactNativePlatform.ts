jest.mock('react-native/Libraries/Utilities/Platform', () => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const fs = require('fs');
  const json = JSON.parse(fs.readFileSync('node_modules/react-native/package.json', 'utf8'));
  const platform = jest.requireActual('react-native/Libraries/Utilities/Platform');
  const version = json.version.toString().split('.');
  return {
    ...platform,
    constants: {
      ...platform.constants,
      reactNativeVersion: {
        major: version[0],
        minor: version[1],
        patch: version[2],
      },
    },
  };
});
