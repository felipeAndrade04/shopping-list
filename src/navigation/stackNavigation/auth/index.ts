export * from './AuthStackNavigation';
export * from './AuthStackNavigation.types';
