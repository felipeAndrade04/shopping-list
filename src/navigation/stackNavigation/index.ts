export * from './auth';
export * from './MainStackNavigation';
