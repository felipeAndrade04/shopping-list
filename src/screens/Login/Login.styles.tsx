import { RectButton } from 'react-native-gesture-handler';
import styled from 'styled-components/native';

export const Container = styled.View`
  flex: 1;
  align-items: center;
  background-color: ${({ theme }) => theme.colors.background};
  padding: 0 24px;
`;

export const LogoContainer = styled.View`
  margin-bottom: 24px;
`;

export const InputsContainer = styled.View`
  width: 100%;
  flex-direction: column;
  align-items: flex-end;
  margin-top: 48px;
`;

export const SimpleButton = styled(RectButton as any)`
  padding: 8px 12px;
  border-radius: 6px;
  margin-bottom: 24px;
`;

export const SimpleButtonBoldText = styled.Text`
  font-size: 14px;
  font-family: ${({ theme }) => theme.fonts.bold};
`;

export const SimpleButtonText = styled.Text`
  font-size: 14px;
`;
