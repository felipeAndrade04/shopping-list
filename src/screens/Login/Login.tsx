import { Button, Input, Message, Spacer } from '@app/components';
import { AuthStackNavigationProps, AuthStackParamList } from '@app/navigation/stackNavigation/auth';
import React, { useState } from 'react';
import { StatusBar } from 'react-native';
import * as S from './Login.styles';

interface Props {
  navigation: AuthStackNavigationProps;
}

export function Login({ navigation }: Props) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');

  function onSubmit() {
    console.log(email, password);
  }

  function onNavigation(name: keyof AuthStackParamList) {
    navigation.navigate(name);
  }

  return (
    <S.Container>
      <StatusBar barStyle={'dark-content'} />
      <Message title="Olá," description="faça login para começar." />

      <S.InputsContainer>
        <Input value={email} onChangeText={setEmail} placeholder="E-mail" />
        <Spacer dimesion={12} />
        <Input value={password} onChangeText={setPassword} placeholder="Senha" />
        <S.SimpleButton>
          <S.SimpleButtonBoldText>Recuperar senha</S.SimpleButtonBoldText>
        </S.SimpleButton>
      </S.InputsContainer>

      <Button onPress={onSubmit} marginTop={24} marginBottom={24}>
        Entrar
      </Button>

      <S.SimpleButton onPress={() => onNavigation('Register')}>
        <S.SimpleButtonText>
          Ainda não possui um cadastro?<S.SimpleButtonBoldText> Criar conta</S.SimpleButtonBoldText>
        </S.SimpleButtonText>
      </S.SimpleButton>
    </S.Container>
  );
}
