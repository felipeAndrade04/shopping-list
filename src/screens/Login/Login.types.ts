import { AuthStackNavigationProps } from '@app/navigation/stackNavigation/auth';

export interface LoginProps {
  navigation: AuthStackNavigationProps;
}

export interface LoginFormData {
  email: string;
  password: string;
}
