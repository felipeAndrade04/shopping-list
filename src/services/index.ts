import { auth, db, storage } from '@app/config';
import { FirebaseAuthentication } from './auth/firebase-authentication';
import { FirebaseShopping } from './shopping/firebase-shopping';
import { FirebaseStorage } from './storage/firebase-storage';

const services = {
  auth: new FirebaseAuthentication(auth),
  shopping: new FirebaseShopping(db),
  storage: new FirebaseStorage(storage),
};

export default services;
