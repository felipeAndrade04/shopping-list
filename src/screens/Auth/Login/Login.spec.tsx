import React from 'react';
import { fireEvent, render, renderHook, waitFor } from '@testing-library/react-native';
import { Login } from './Login';
import { AuthStackNavigationProps } from '@app/navigation/stackNavigation/auth';
import services from '@app/services';
import { LoginParams } from '@app/services/auth';
import { useAuth } from '@app/hooks';
import { resetStore, wrapper } from '@app/utils/testUtils';
import '@app/mocks/MockReactNativePlatform';

describe('Login page', () => {
  const navigation = {
    ...jest.requireMock('@app/navigation/stackNavigation/auth'),
    navigate: jest.fn(),
  } as AuthStackNavigationProps;

  afterEach(() => {
    jest.resetAllMocks();
    resetStore();
  });

  it('Should render correctly', () => {
    const tree = render(<Login navigation={navigation}></Login>, { wrapper }).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should be required email and password', async () => {
    const { queryAllByText, queryByText } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });

    const button = queryByText(/entrar/i);

    await waitFor(() => {
      fireEvent.press(button);
    });

    expect(queryAllByText(/campo obrigatório/i)).toHaveLength(2);
  });

  it('Should be required email', async () => {
    const { queryByText, queryByTestId } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });

    const button = queryByText(/entrar/i);
    const inputPassword = queryByTestId(/input-password/i);

    await waitFor(() => {
      fireEvent.changeText(inputPassword, '12345678');
      fireEvent.press(button);
    });

    expect(queryByText(/campo obrigatório/i)).toBeTruthy();
  });

  it('Should be required password', async () => {
    const { queryByText, queryByTestId } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });

    const button = queryByText(/entrar/i);
    const inputEmail = queryByTestId(/input-email/i);

    await waitFor(() => {
      fireEvent.changeText(inputEmail, 'teste@email.com');
      fireEvent.press(button);
    });

    expect(queryByText(/campo obrigatório/i)).toBeTruthy();
  });

  it('Should be valid email and password', async () => {
    const { queryByText, queryByTestId } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });

    const button = queryByText(/entrar/i);
    const inputEmail = queryByTestId(/input-email/i);
    const inputPassword = queryByTestId(/input-password/i);

    await waitFor(() => {
      fireEvent.changeText(inputEmail, 'teste');
      fireEvent.changeText(inputPassword, '1234');
      fireEvent.press(button);
    });

    expect(queryByText(/informe um email válido/i)).toBeTruthy();
    expect(queryByText(/a senha precisa de 8 caracteres/i)).toBeTruthy();
  });

  it('Should be valid email', async () => {
    const { queryByText, queryByTestId } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });

    const button = queryByText(/entrar/i);
    const inputEmail = queryByTestId(/input-email/i);
    const inputPassword = queryByTestId(/input-password/i);

    await waitFor(() => {
      fireEvent.changeText(inputEmail, 'teste');
      fireEvent.changeText(inputPassword, '12345678');
      fireEvent.press(button);
    });

    expect(queryByText(/informe um email válido/i)).toBeTruthy();
  });

  it('Should be valid password', async () => {
    const { queryByText, queryByTestId } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });

    const button = queryByText(/entrar/i);
    const inputEmail = queryByTestId(/input-email/i);
    const inputPassword = queryByTestId(/input-password/i);

    await waitFor(() => {
      fireEvent.changeText(inputEmail, 'teste@email.com');
      fireEvent.changeText(inputPassword, '1234');
      fireEvent.press(button);
    });

    expect(queryByText(/a senha precisa de 8 caracteres/i)).toBeTruthy();
  });

  it('Should be do login correctly', async () => {
    const user = {
      email: 'teste@email.com',
      password: '12345678',
    };
    const userName = 'teste';
    const userId = '1';
    jest.spyOn(services.auth, 'login').mockImplementation((user: LoginParams) => {
      return new Promise((resolve) => {
        resolve({
          id: userId,
          name: userName,
          email: user.email,
        });
      });
    });
    const { queryByText, queryByTestId } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });
    const { result } = renderHook(() => useAuth(), { wrapper });

    const button = queryByText(/entrar/i);
    const inputEmail = queryByTestId(/input-email/i);
    const inputPassword = queryByTestId(/input-password/i);

    await waitFor(() => {
      fireEvent.changeText(inputEmail, user.email);
      fireEvent.changeText(inputPassword, user.password);
      fireEvent.press(button);
    });

    expect(services.auth.login).toBeCalledWith(user);
    expect(result.current.user.email).toBe(user.email);
    expect(result.current.user.name).toBe(userName);
    expect(result.current.user.id).toBe(userId);
    expect(result.current.isAuthenticated).toBeTruthy();
  });

  it('Should failed login', async () => {
    const user = {
      email: 'teste@email.com',
      password: '12345678',
    };
    jest.spyOn(services.auth, 'login').mockImplementation(() => {
      return new Promise((_resolve, reject) => {
        return reject(new Error('Login failed'));
      });
    });
    const { queryByText, queryByTestId } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });
    const { result } = renderHook(() => useAuth(), { wrapper });

    const button = queryByText(/entrar/i);
    const inputEmail = queryByTestId(/input-email/i);
    const inputPassword = queryByTestId(/input-password/i);

    await waitFor(() => {
      fireEvent.changeText(inputEmail, user.email);
      fireEvent.changeText(inputPassword, user.password);
      fireEvent.press(button);
    });

    expect(services.auth.login).toBeCalled();
    expect(services.auth.login).rejects.toThrow();
    expect(result.current.isAuthenticated).toBeFalsy();
    expect(result.current.user).toBeNull();
  });

  it('Should navigate to forgot password page', () => {
    const { queryByText } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });

    const button = queryByText(/recuperar senha/i);
    fireEvent.press(button);

    expect(navigation.navigate).toHaveBeenCalledWith('ForgotPassword');
  });

  it('Should navigate to register', () => {
    const { queryByTestId } = render(<Login navigation={navigation}></Login>, {
      wrapper,
    });

    const button = queryByTestId(/register/i);
    fireEvent.press(button);

    expect(navigation.navigate).toHaveBeenCalledWith('Register');
  });
});
