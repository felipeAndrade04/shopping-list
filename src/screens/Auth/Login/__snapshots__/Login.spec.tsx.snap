// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page Should render correctly 1`] = `
<RCTScrollView
  automaticallyAdjustContentInsets={false}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  contentInset={
    Object {
      "bottom": 0,
    }
  }
  enableAutomaticScroll={true}
  enableOnAndroid={false}
  enableResetScrollToCoords={true}
  extraHeight={75}
  extraScrollHeight={0}
  getScrollResponder={[Function]}
  handleOnScroll={[Function]}
  keyboardDismissMode="interactive"
  keyboardOpeningTime={250}
  keyboardSpace={0}
  onScroll={[Function]}
  resetKeyboardSpace={[Function]}
  scrollEventThrottle={1}
  scrollForExtraHeightOnAndroid={[Function]}
  scrollIntoView={[Function]}
  scrollToEnd={[Function]}
  scrollToFocusedInput={[Function]}
  scrollToPosition={[Function]}
  showsVerticalScrollIndicator={true}
  update={[Function]}
  viewIsInsideTabBar={false}
>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#EEEEEE",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "paddingBottom": 0,
            "paddingLeft": 24,
            "paddingRight": 24,
            "paddingTop": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "marginBottom": 24,
            },
          ]
        }
      >
        <logo.svg />
      </View>
      <View
        style={
          Array [
            Object {
              "justifyContent": "center",
              "width": "100%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#0D0A0B",
                "fontFamily": "Poppins-Medium",
                "fontSize": 24,
              },
            ]
          }
        >
          Olá,
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#0D0A0B",
                "fontFamily": "Poppins-Regular",
                "fontSize": 16,
                "marginTop": 14,
              },
            ]
          }
        >
          faça login para começar.
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-end",
              "flexDirection": "column",
              "marginTop": 48,
              "width": "100%",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "width": "100%",
              },
            ]
          }
        >
          <View
            isFocused={false}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#FFFFFF",
                  "borderRadius": 6,
                  "borderStyle": "solid",
                  "borderWidth": 2,
                  "flexDirection": "row",
                  "opacity": 1,
                  "paddingBottom": 0,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "paddingTop": 0,
                  "transition": "border-color 0.3s ease-in",
                  "width": "100%",
                },
              ]
            }
          >
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="E-mail"
              style={
                Array [
                  Object {
                    "color": "#0D0A0B",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontFamily": "Poppins-Regular",
                    "fontSize": 16,
                    "height": 50,
                    "paddingBottom": 8,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 8,
                  },
                ]
              }
              testID="input-email"
              value=""
            />
          </View>
        </View>
        <View
          dimesion={12}
          style={
            Array [
              Object {
                "height": 12,
                "width": 12,
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "width": "100%",
              },
            ]
          }
        >
          <View
            isFocused={false}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#FFFFFF",
                  "borderRadius": 6,
                  "borderStyle": "solid",
                  "borderWidth": 2,
                  "flexDirection": "row",
                  "opacity": 1,
                  "paddingBottom": 0,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "paddingTop": 0,
                  "transition": "border-color 0.3s ease-in",
                  "width": "100%",
                },
              ]
            }
          >
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Senha"
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "color": "#0D0A0B",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontFamily": "Poppins-Regular",
                    "fontSize": 16,
                    "height": 50,
                    "paddingBottom": 8,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                    "paddingTop": 8,
                  },
                ]
              }
              testID="input-password"
              value=""
            />
            <Text />
          </View>
        </View>
        <RNGestureHandlerButton
          activeOpacity={0.105}
          collapsable={false}
          delayLongPress={600}
          handlerTag={1}
          handlerType="NativeViewGestureHandler"
          onActiveStateChange={[Function]}
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          onPress={[Function]}
          style={
            Object {
              "borderRadius": 6,
              "marginBottom": 24,
              "paddingBottom": 8,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 8,
            }
          }
          underlayColor="black"
        >
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "black",
                "borderBottomLeftRadius": undefined,
                "borderBottomRightRadius": undefined,
                "borderRadius": 6,
                "borderTopLeftRadius": undefined,
                "borderTopRightRadius": undefined,
                "bottom": 0,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins-Bold",
                  "fontSize": 14,
                },
              ]
            }
          >
            Recuperar senha
          </Text>
        </RNGestureHandlerButton>
      </View>
      <RNGestureHandlerButton
        activeOpacity={0.105}
        collapsable={false}
        color="#009Cf7"
        delayLongPress={600}
        enabled={false}
        handlerTag={2}
        handlerType="NativeViewGestureHandler"
        isLoading={false}
        marginBottom={24}
        marginTop={24}
        onActiveStateChange={[Function]}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#009Cf7",
            "borderColor": "black",
            "borderRadius": 6,
            "borderStyle": "solid",
            "borderWidth": 0,
            "display": "flex",
            "height": 56,
            "justifyContent": "center",
            "marginBottom": 24,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 24,
            "opacity": 0.6,
            "paddingBottom": 8,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 8,
            "width": "100%",
          }
        }
        underlayColor="black"
        variant="solid"
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "black",
              "borderBottomLeftRadius": undefined,
              "borderBottomRightRadius": undefined,
              "borderRadius": 6,
              "borderTopLeftRadius": undefined,
              "borderTopRightRadius": undefined,
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <Text
          color="#009Cf7"
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "Poppins-Bold",
                "fontSize": 18,
              },
            ]
          }
          variant="solid"
        >
          Entrar
        </Text>
      </RNGestureHandlerButton>
      <RNGestureHandlerButton
        activeOpacity={0.105}
        collapsable={false}
        delayLongPress={600}
        handlerTag={3}
        handlerType="NativeViewGestureHandler"
        onActiveStateChange={[Function]}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        onPress={[Function]}
        style={
          Object {
            "borderRadius": 6,
            "marginBottom": 24,
            "paddingBottom": 8,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 8,
          }
        }
        testID="register"
        underlayColor="black"
      >
        <View
          collapsable={false}
          style={
            Object {
              "backgroundColor": "black",
              "borderBottomLeftRadius": undefined,
              "borderBottomRightRadius": undefined,
              "borderRadius": 6,
              "borderTopLeftRadius": undefined,
              "borderTopRightRadius": undefined,
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Poppins-Regular",
                "fontSize": 14,
              },
            ]
          }
        >
          Ainda não possui um cadastro?
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Poppins-Bold",
                  "fontSize": 14,
                },
              ]
            }
          >
             Criar conta
          </Text>
        </Text>
      </RNGestureHandlerButton>
    </View>
  </View>
</RCTScrollView>
`;
